fastcrypto/src/groups/secp256k1.rs

```diff
@@ -0,0 +1,215 @@
+use ark_ec::{CurveGroup, Group, VariableBaseMSM};
+use ark_ff::{Field, One, UniformRand, Zero};
+use ark_secp256k1::{Affine, Fq, Fr, Projective};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use derive_more::{Add, From, Neg, Sub};
+use fastcrypto_derive::GroupOpsExtend;
+use k256::elliptic_curve::bigint::{ArrayDecoding, ArrayEncoding};
+use k256::elliptic_curve::hash2curve::{ExpandMsgXmd, GroupDigest};
+use k256::elliptic_curve::sec1::ToEncodedPoint;
+use k256::elliptic_curve::Group as GroupTrait;
+use k256::Secp256k1;
+use serde::{de, Deserialize};
+use std::ops::{Div, Mul};
+
+/// Size of a serialized scalar in bytes.
+pub const SCALAR_SIZE_IN_BYTES: usize = 32;
+
+/// Size of a serialized point in bytes. This uses compressed serialization.
+pub const POINT_SIZE_IN_BYTES: usize = 33;
+
+/// A point on the Secp256k1 curve in projective coordinates.
+#[derive(Default, Clone, Copy, Debug, PartialEq, Eq, From, Add, Sub, Neg, GroupOpsExtend)]
+pub struct ProjectivePoint(pub(crate) Projective);
+
+impl GroupElement for ProjectivePoint {
+    type ScalarType = Scalar;
+
+    fn zero() -> Self {
+        Self(Projective::zero())
+    }
+
+    fn generator() -> Self {
+        Self(Projective::generator())
+    }
+}
+
+impl Doubling for ProjectivePoint {
+    fn double(self) -> Self {
+        ProjectivePoint::from(self.0.double())
+    }
+}
+
+impl Mul<Scalar> for ProjectivePoint {
+    type Output = ProjectivePoint;
+
+    fn mul(self, rhs: Scalar) -> ProjectivePoint {
+        ProjectivePoint::from(self.0 * rhs.0)
+    }
+}
+
+impl Div<Scalar> for ProjectivePoint {
+    type Output = Result<ProjectivePoint, FastCryptoError>;
+
+    #[allow(clippy::suspicious_arithmetic_impl)]
+    fn div(self, rhs: Scalar) -> Result<ProjectivePoint, FastCryptoError> {
+        Ok(self * rhs.inverse()?)
+    }
+}
+
+impl ToFromByteArray<POINT_SIZE_IN_BYTES> for ProjectivePoint {
+    fn from_byte_array(bytes: &[u8; POINT_SIZE_IN_BYTES]) -> Result<Self, FastCryptoError> {
+        Ok(ProjectivePoint(
+            Projective::deserialize_compressed(bytes.as_slice())
+                .map_err(|_| FastCryptoError::InvalidInput)?,
+        ))
+    }
+
+    fn to_byte_array(&self) -> [u8; POINT_SIZE_IN_BYTES] {
+        let mut bytes = [0u8; POINT_SIZE_IN_BYTES];
+        self.0
+            .serialize_compressed(&mut bytes[..])
+            .expect("Is always 33 bytes");
+        bytes
+    }
+}
+
+serialize_deserialize_with_to_from_byte_array!(ProjectivePoint);
+
+impl MultiScalarMul for ProjectivePoint {
+    fn multi_scalar_mul(
+        scalars: &[Self::ScalarType],
+        points: &[Self],
+    ) -> Result<Self, FastCryptoError> {
+        // Fail early if the lengths do not match
+        if scalars.len() != points.len() {
+            return Err(FastCryptoError::InvalidInput);
+        }
+        let scalars = scalars.iter().map(|s| s.0).collect::<Vec<_>>();
+        let points = points.iter().map(|g| g.0.into_affine()).collect::<Vec<_>>();
```

Comment:
why is this conversion needed?

Reply:
It's because `Projective::msm` (a bit confusingly, and redundant imo) takes a list of points in affine form as input. I found a batch conversion method, but I don't think the conversion can be avoided.